File: problem/Best_Time_Buy_Sell/maxProfit.py
class Solution:
    def maxProfit(self, prices, strat, k):
        n = len(prices)
        h = k // 2

        sp = [0] * n
        base = 0
        for i in range(n):
            sp[i] = strat[i] * prices[i]
            base += sp[i]

        if n == k:
            win_orig = base
            win_mod = 0
            for i in range(h, n):
                win_mod += prices[i]
            change = win_mod - win_orig
            return base + (change if change > 0 else 0)

        win_orig = 0
        for i in range(k):
            win_orig += sp[i]

        win_mod = 0
        for i in range(h, k):
            win_mod += prices[i]

        max_ch = win_mod - win_orig

        for i in range(1, n - k + 1):
            win_orig += sp[i + k - 1] - sp[i - 1]
            win_mod += prices[i + k - 1] - prices[i - 1 + h]
            diff = win_mod - win_orig
            if diff > max_ch:
                max_ch = diff

        return base + (max_ch if max_ch > 0 else 0)

File: problem/Best_Time_Buy_Sell/maxProfit.c
long long maxProfit(int *prices, int pricesSize, int *strat, int stratSize, int k)
{
    int n = pricesSize;
    int h = k / 2;

    long long *sp = (long long *)malloc(sizeof(long long) * n);
    long long base = 0;
    for (int i = 0; i < n; i++)
    {
        sp[i] = (long long)strat[i] * (long long)prices[i];
        base += sp[i];
    }

    if (n == k)
    {
        long long win_orig = base;
        long long win_mod = 0;
        for (int i = h; i < n; i++)
            win_mod += (long long)prices[i];
        long long change = win_mod - win_orig;
        long long ans = base + (change > 0 ? change : 0);
        free(sp);
        return ans;
    }

    long long win_orig = 0;
    for (int i = 0; i < k; i++)
        win_orig += sp[i];

    long long win_mod = 0;
    for (int i = h; i < k; i++)
        win_mod += (long long)prices[i];

    long long max_ch = win_mod - win_orig;

    for (int i = 1; i <= n - k; i++)
    {
        win_orig += sp[i + k - 1] - sp[i - 1];
        win_mod += (long long)prices[i + k - 1] - (long long)prices[i - 1 + h];
        long long diff = win_mod - win_orig;
        if (diff > max_ch)
            max_ch = diff;
    }

    long long ans = base + (max_ch > 0 ? max_ch : 0);
    free(sp);
    return ans;
}